src/Groups/directproducts.jl

```diff
@@ -0,0 +1,486 @@
+   if length(L)==0 throw(ArgumentError("Empty list")) end
+   l=collect(L)
+   @assert all(x -> parent(x) == parent(l[1]), l)
+   return sub(parent(l[1]),l)
+end
+
+function sub(L::Vector{GAPGroupElem{DirectProductOfGroups}})
+   if length(L)==0 throw(ArgumentError("Empty list")) end
+   l=collect(L)
+   @assert all(x -> parent(x) == parent(l[1]), l)
+   return sub(parent(l[1]),l)
+end
+
+function Base.show(io::IO, G::DirectProductOfGroups)
+   if G.isfull
+      print(io, "DirectProduct of ")
+      display(G.L)
+   else
+      print(io, GAP.gap_to_julia(GAP.Globals.StringViewObj(G.X)))
+   end
+end
+
+#=
+function Base.show(io::IO, x::GAPGroupElem{DirectProductOfGroups{S,T}}) where { S, T }
+   print(io, "DirectProductElement( ", GAP.gap_to_julia(GAP.Globals.StringViewObj(projection(parent(x),1)(x).X)), " , ", GAP.gap_to_julia(GAP.Globals.StringViewObj(projection(parent(x),2)(x).X)), " )")
+end
+=#
+
+# if a subgroup of a direct product of groups is also a direct product of groups
+"""
+    write_as_full(G::DirectProductOfGroups)
+If `G` is a subgroup of the direct product `G_1 x ... x G_n` such that `G = H_1 x ... x H_n` for `H_i` subgroup of `G_i`, return `G` as full direct product of the `H_i`. If such `H_i` do not exist, an ERROR is returned.
+"""
+function write_as_full(G::DirectProductOfGroups)
+   if G.isfull
+      return G
+   else
+      LK = [image(projection(G,j))[1] for j in 1:G.n]
+      H = direct_product(LK)
+# index(H,G)==1 does not work because it does not recognize G as a subgroup of H
+      if order(H)==order(G)
+         return H
+      else
+         throw(ArgumentError("G is not a direct product of groups"))
+      end
+   end
+end
+
+"""
+    isfull_direct_product(G::DirectProductOfGroups)
+Return whether `G` is direct product of its factors (`false` if it is a proper subgroup).
+"""
+isfull_direct_product(G::DirectProductOfGroups) = G.isfull
+
+Base.:^(H::DirectProductOfGroups, y::GAPGroupElem) = sub([h^y for h in gens(H)])[1]
+
+
+################################################################################
+#
+#  Semidirect products
+#  
+################################################################################
+
+"""
+    semidirect_product(N::S, f::GAPGroupHomomorphism, H::T)
+Return the semidirect product of `N` and `H`, of type ``SemidirectProductOfGroups{S,T}``, where `f` is a homomorphism from `H` to `Aut`(`N`).
+"""
+function semidirect_product(N::S, f::GAPGroupHomomorphism{T,AutomorphismGroup{S}}, H::T) where S <: GAPGroup where T <: GAPGroup
+   sdp=GAP.Globals.SemidirectProduct(H.X,f.map,N.X)
+   return SemidirectProductOfGroups(sdp,N,H,f,sdp,true)
+end
+
+# return the element (a,b) in G
+function (G::SemidirectProductOfGroups{S,T})(a::GAPGroupElem{S},b::GAPGroupElem{T}) where {S,T}
+# simply put parent(L[d+1])==W.H does not work. Example: if I want to write explicitly a permutation in H proper subgroup of Sym(n).
+   Gfull = SemidirectProductOfGroups(G.Xfull,G.N,G.H,G.f,G.Xfull,true)
```

Comment:
This looks as if you want to construct the parent of `G`. So that would suggest this code:
```suggestion
   Gfull = parent(G)
```
Of course you then also need to provide a corresponding `parent`  method. The same holds in several other places.

That said, it is super wasteful (i.e.: leads to very slow code using much more memory than necessary) to construct the full parent object here, just to call embedding on it -- when you could simply write `g = a.X^GAP.Globals.Embedding(G.Xfull, 1) * b.X^GAP.Globals.Embedding(G.Xfull, 2)` and then check if `g in G.X` in the `@assert` below. Of course the `return` then also needs to be adjusted.

And this remarks applies to multiple other places in this PR, too!

Reply:
Isn't `parent` confusing? The method `parent` is used for elements, so if `G` is parent of `x`, it means that `x` is an element of `G`, not a subset.
Anyway, we can add a field to the type `DirectProductGroup` (and all the others) storing the `parent` object of the group `G`, that is what I called the full direct product. This will save time when we need the parent of `G`, but it will load the type with a further heavy field (the covering group). Is it the case we do it?
PS: a similar field makes sense also for groups of type `FPGroup` and `PermGroup`.

Reply:
I am fine with calling it something else, I mentioned `parent` because that's how it is called in GAP; but yeah, in OSCAR so far, it seems `parent` is for mapping from elements to their containing algebraic structure. For modules, it seems `supermodule` is used for this purpose, so perhaps we should call it `supergroup`?

But anyway, in the situation we discuss right now, this is not needed at all, as I pointed out, as you are storing Xfull anyway. So let's focus on that first, and discuss the question of introducing something like a "supergroup" (and what it should do) another time.